File: Quiz_5/Tanvi_Kapoor_590016403/Quiz_5_Debugging_Question_1.c
#include <stdio.h>

void findUniquePair(int arr[], int n, int* num1, int* num2) {
    int xor_sum = 0;
    
    // Step 1: XOR all elements
    for (int i = 0; i < n; i++) {
        xor_sum ^= arr[i];
    }

    // Step 2: Isolate the rightmost set bit
    int diff_bit = xor_sum & -xor_sum;

    *num1 = 0;
    *num2 = 0;

    // Step 3: Divide into two groups and XOR separately
    for (int i = 0; i < n; i++) {
        if (arr[i] & diff_bit)
            *num1 ^= arr[i];
        else
            *num2 ^= arr[i];
    }
}

int main() {
    int arr[] = {2, 4, 7, 9, 2, 4};
    int n = sizeof(arr) / sizeof(arr[0]);
    int num1, num2;

    findUniquePair(arr, n, &num1, &num2);

    printf("The two unique numbers are: %d and %d\n", num1, num2);

    return 0;
}

File: Quiz_5/Tanvi_Kapoor_590016403/Quiz_5_Debugging_Question_1.py
def find_unique_pair(arr):
    xor_sum = 0

    # Step 1: XOR all elements
    for num in arr:
        xor_sum ^= num

    # Step 2: Isolate the rightmost set bit (this bit differs in the two unique numbers)
    diff_bit = xor_sum & -xor_sum

    unique1 = 0
    unique2 = 0

    # Step 3: Divide numbers into two groups based on the isolated bit
    for num in arr:
        if num & diff_bit:
            unique1 ^= num
        else:
            unique2 ^= num

    return unique1, unique2


# ✅ Example usage:
arr = [2, 4, 7, 9, 2, 4]
u1, u2 = find_unique_pair(arr)
print("The two unique numbers are:", u1, "and", u2)
